File: OpenCount/Models/CountFormula.swift
import Foundation

// MARK: - CountFormula

final class CountFormula: ObservableObject, Identifiable, Codable {
    var id: UUID
    var name: String
    var expression: String
    var unit: String
    var sortOrder: Int
    weak var session: CountSession?

    init(
        id: UUID = UUID(),
        name: String,
        expression: String,
        unit: String = "",
        sortOrder: Int = 0,
        session: CountSession? = nil
    ) {
        self.id = id
        self.name = name
        self.expression = expression
        self.unit = unit
        self.sortOrder = sortOrder
        self.session = session
    }

    enum CodingKeys: String, CodingKey {
        case id, name, expression, unit, sortOrder
    }

    required init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id         = try c.decode(UUID.self,   forKey: .id)
        name       = try c.decode(String.self, forKey: .name)
        expression = try c.decode(String.self, forKey: .expression)
        unit       = try c.decode(String.self, forKey: .unit)
        sortOrder  = try c.decode(Int.self,    forKey: .sortOrder)
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id,        forKey: .id)
        try c.encode(name,      forKey: .name)
        try c.encode(expression, forKey: .expression)
        try c.encode(unit,      forKey: .unit)
        try c.encode(sortOrder, forKey: .sortOrder)
    }
}

// MARK: - FormulaEvaluator

struct FormulaEvaluator {
    static func evaluate(formula: CountFormula, tally: [String: Int]) -> Double? {
        var expression = formula.expression
        let sortedNames = tally.keys.sorted { $0.count > $1.count }
        for name in sortedNames {
            let count = tally[name] ?? 0
            let pattern = NSRegularExpression.escapedPattern(for: name)
            if let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) {
                expression = regex.stringByReplacingMatches(
                    in: expression,
                    range: NSRange(expression.startIndex..., in: expression),
                    withTemplate: "\(count)"
                )
            }
        }
        return evaluateArithmetic(expression)
    }

    private static func evaluateArithmetic(_ expression: String) -> Double? {
        let cleaned = expression.replacingOccurrences(of: " ", with: "")
        var index = cleaned.startIndex
        return parseExpression(cleaned, index: &index)
    }

    private static func parseExpression(_ s: String, index: inout String.Index) -> Double? {
        var result = parseTerm(s, index: &index)
        while index < s.endIndex {
            let op = s[index]
            guard op == "+" || op == "-" else { break }
            s.formIndex(after: &index)
            guard let rhs = parseTerm(s, index: &index) else { return nil }
            result = op == "+" ? (result ?? 0) + rhs : (result ?? 0) - rhs
        }
        return result
    }

    private static func parseTerm(_ s: String, index: inout String.Index) -> Double? {
        var result = parseFactor(s, index: &index)
        while index < s.endIndex {
            let op = s[index]
            guard op == "*" || op == "/" else { break }
            s.formIndex(after: &index)
            guard let rhs = parseFactor(s, index: &index) else { return nil }
            if op == "*" { result = (result ?? 0) * rhs }
            else { guard rhs != 0 else { return nil }; result = (result ?? 0) / rhs }
        }
        return result
    }

    private static func parseFactor(_ s: String, index: inout String.Index) -> Double? {
        guard index < s.endIndex else { return nil }
        if s[index] == "(" {
            s.formIndex(after: &index)
            let result = parseExpression(s, index: &index)
            if index < s.endIndex && s[index] == ")" { s.formIndex(after: &index) }
            return result
        }
        if s[index] == "-" {
            s.formIndex(after: &index)
            return parseFactor(s, index: &index).map { -$0 }
        }
        var numStr = ""
        while index < s.endIndex && (s[index].isNumber || s[index] == ".") {
            numStr.append(s[index]); s.formIndex(after: &index)
        }
        return Double(numStr)
    }
}

File: packages/shared/src/commonMain/kotlin/com/opencount/shared/model/CountFormula.kt
package com.opencount.shared.model

import kotlinx.serialization.Serializable

@Serializable
data class CountFormula(
    val id: String,
    val name: String,
    val expression: String,
    val unit: String = "",
    val sortOrder: Int = 0,
)

object FormulaEvaluator {
    fun evaluate(formula: CountFormula, tally: Map<String, Int>): Double? {
        var expression = formula.expression
        val sortedNames = tally.keys.sortedByDescending { it.length }
        for (name in sortedNames) {
            val count = tally[name] ?: 0
            expression = expression.replace(name, count.toString(), ignoreCase = true)
        }
        return evaluateArithmetic(expression)
    }

    private fun evaluateArithmetic(expression: String): Double? {
        val cleaned = expression.replace(" ", "")
        val tokens = tokenize(cleaned) ?: return null
        return parseExpression(tokens)
    }

    private fun tokenize(s: String): List<String>? {
        val tokens = mutableListOf<String>()
        var i = 0
        while (i < s.length) {
            val c = s[i]
            when {
                c.isDigit() || c == '.' -> {
                    val start = i
                    while (i < s.length && (s[i].isDigit() || s[i] == '.')) i++
                    tokens.add(s.substring(start, i))
                }
                c in "+-*/()" -> {
                    tokens.add(c.toString())
                    i++
                }
                else -> return null
            }
        }
        return tokens
    }

    private var pos = 0
    private lateinit var tokens: List<String>

    private fun parseExpression(tokens: List<String>): Double? {
        this.tokens = tokens; pos = 0
        return parseExpr()
    }

    private fun parseExpr(): Double? {
        var result = parseTerm() ?: return null
        while (pos < tokens.size && (tokens[pos] == "+" || tokens[pos] == "-")) {
            val op = tokens[pos++]
            val rhs = parseTerm() ?: return null
            result = if (op == "+") result + rhs else result - rhs
        }
        return result
    }

    private fun parseTerm(): Double? {
        var result = parseFactor() ?: return null
        while (pos < tokens.size && (tokens[pos] == "*" || tokens[pos] == "/")) {
            val op = tokens[pos++]
            val rhs = parseFactor() ?: return null
            result = if (op == "*") result * rhs else if (rhs != 0.0) result / rhs else return null
        }
        return result
    }

    private fun parseFactor(): Double? {
        if (pos >= tokens.size) return null
        return when (val token = tokens[pos++]) {
            "(" -> {
                val result = parseExpr()
                if (pos < tokens.size && tokens[pos] == ")") pos++ else return null
                result
            }
            "-" -> parseFactor()?.let { -it }
            else -> token.toDoubleOrNull()
        }
    }
}
